components/script/layout_wrapper.rs

```diff
@@ -234,12 +242,23 @@ impl<'ln> TNode for ServoLayoutNode<'ln> {
         old_value - 1
     }
 
-    fn borrow_data(&self) -> Option<AtomicRef<PersistentStyleData>> {
-        self.get_style_data().map(|d| d.borrow())
+    fn get_existing_style(&self) -> Option<Arc<ComputedValues>> {
+        self.borrow_data().and_then(|x| x.style.clone())
     }
 
-    fn mutate_data(&self) -> Option<AtomicRefMut<PersistentStyleData>> {
-        self.get_style_data().map(|d| d.borrow_mut())
+    fn set_style(&self, style: Option<Arc<ComputedValues>>) {
+        self.mutate_data().unwrap().style = style;
+    }
+
+    fn take_pseudo_styles(&self) -> PseudoStyles {
+        use std::mem;
+        let mut tmp = PseudoStyles::default();
+        mem::swap(&mut tmp, &mut self.mutate_data().unwrap().per_pseudo);
+        tmp
+    }
+
+    fn set_pseudo_styles(&self, styles: PseudoStyles) {
+        self.mutate_data().unwrap().per_pseudo = styles;
```

Comment:
> debug_assert! that the current pseudo styles is empty?

Done.

I considered the Option thing, but I think it makes the semantics more confusing, and empty hashmap construction is cheap (there's no allocation). So I don't think it's worth it.

Converting the HashMap to a vec is on my TODO list (in the googledoc), but I'll do it once the architecture settles more. I'll have to change this pseudo stuff around a bit anyway to handle the fact that we may not have the hashmap in the gecko case.